src/interpreter.py

```diff
@@ -0,0 +1,108 @@
+
+from objects import *
+
+class Binding:
+  def __init__(self, outer = None):
+    self.bindings = {}
+    self.outer = outer
+
+  def push(self):
+    return Binding(self)
+
+  def pop(self):
+    return self.outer
+
+  def set_variable(self, name, value):
+    self.bindings[name] = value
+
+  def get_value(self, name):
+    print(self.bindings)
+    if name in self.bindings:
+      return self.bindings[name]
+    
+    if self.outer:
+      return self.outer.get_value(name)
+
+    raise Exception(f"Variable '{name}' is not defined")
+
+
+class Interpreter:
+  def __init__(self):
+    # interpreter-wide binding
+    # using this method because I find it easier and clearer than passing the binding
+    # into every function that needs it
```

Comment:
I understand your point, but for some reason my spidey-sense is tingling. I can't see anything wrong with maintaining the binding tree using local variables in `evaluate_thunk`, but somewhere in the back of my brain  a worry is struggling to be heard.

It might be to do with thread safety, in which case it's not really a worry. 

See if you can think of a downside. If not, I'm not going to grade it down.